lib/fss.ex

```diff
@@ -13,19 +13,99 @@ defmodule FSS do
 
   defmodule S3 do
     defmodule Config do
-      defstruct [:region, :access_key_id, :secret_access_key]
+      defstruct [
+        :access_key_id,
+        :region,
+        :secret_access_key,
+        {:endpoint, "amazonaws.com"},
+        :token
+      ]
 
       @type t :: %__MODULE__{
-              region: String.t(),
               access_key_id: String.t(),
-              secret_access_key: String.t()
+              endpoint: String.t(),
+              region: String.t(),
+              secret_access_key: String.t(),
+              token: String.t() | nil
             }
+
+      def from_system_env() do
+        %__MODULE__{
+          access_key_id: System.get_env("AWS_ACCESS_KEY_ID"),
+          secret_access_key: System.get_env("AWS_SECRET_ACCESS_KEY"),
+          region: System.get_env("AWS_REGION", System.get_env("AWS_DEFAULT_REGION")),
+          token: System.get_env("AWS_SESSION_TOKEN")
+        }
+      end
     end
 
     defmodule Entry do
-      defstruct [:url, :config]
+      defstruct [:bucket, :key, :port, :config]
+
+      @type t :: %__MODULE__{
+              bucket: String.t(),
+              key: String.t(),
+              port: pos_integer(),
+              config: Config.t()
+            }
+
+      def parse(url, opts \\ []) do
+        opts = Keyword.validate!(opts, config: FSS.S3.Config.from_system_env())
+        uri = URI.parse(url)
+
+        cond do
+          uri.scheme == "s3" ->
+            case uri do
+              %{host: host, path: "/" <> key} when is_binary(host) ->
+                {:ok, %__MODULE__{bucket: host, key: key, port: uri.port, config: opts[:config]}}
+
+              %{host: nil} ->
+                {:error, "host is required"}
+
+              %{path: nil} ->
+                {:error, "path to the resource is required"}
+            end
+
+          uri.scheme == "http" or uri.scheme == "https" ->
+            parts =
+              case String.split(uri.host, ".") do
+                ["s3", region | tail] ->
```

Comment:
So this is specific to AWS, is it specific to AWS on Polars side, or could we support other compatible storages?

Reply:
Good point. We use Minio and it'd be great to be able to use it with this.

Reply:
Looks like Polars itself only supports the s3://-style URLs without specifying the region in the URL (source and the related CloudLocation and CloudType structs). (You can pass the region as a CloudOption though).

The 'virtual-hosted style' URLs seem to be AWS-specific as far as I can find. Some other storages (including Minio) have opt-in support for them, though the exact format of the resulting URLs might be too different to cover with this particular pattern-match.

Reply:
Since Polars only supports the `s3://` version - thanks for checking that, @Qqwy ! -, I'm going to remove the handle of `http(s)` URLs for now.